src/main/kotlin/org/rust/lang/core/resolve/NameResolution.kt

```diff
@@ -734,52 +765,59 @@ private fun reexportedMacros(item: RsExternCrateItem): List<RsMacro>? {
-            val path = useSpeck.path
-            if (path != null) {
-                consumer(useSpeck, path)
+        for ((crateName, macroName) in collect2segmentPaths(root)) {
+            val crateRoot = exportingMacrosCrates[crateName] as? RsFile
+                ?: findDependencyCrateByName(useItem, crateName)
+                ?: continue
+            val exportedMacros = exportedMacros(crateRoot)
+            addAll(if (macroName == null) {
+                exportedMacros
             } else {
-                useSpeck.useGroup?.useSpeckList?.forEach { forEachRootUseSpeck(it, consumer) }
-            }
+                exportedMacros.filter { it.name == macroName }
+            })
         }
+    }
+}
 
-        forEachRootUseSpeck(root) { speck, path ->
-            val basePath = path.basePath()
-            // If path name is not in `exportingMacrosCrates` then
-            // it can't be resolved to extern crate mod exporting macros.
-            // Note, it significantly improves performance of macro resolution
-            val expectedMod = exportingMacrosCrates[basePath.referenceName] ?: return@forEachRootUseSpeck
-
-            val mod = basePath.reference.resolve() as? RsFile ?: return@forEachRootUseSpeck
-            // Don't try to resolve leaf items if `mod` can't bring macros at all
-            if (mod == expectedMod) {
-                val crateExportedMacros = lazy { exportedMacros(mod) }
-
-                speck.forEachLeafSpeck { leafSpeck ->
-                    // We should use multi resolve here
-                    // because single use speck can import module and macro at same time
-                    val resolvedItems = leafSpeck.path?.reference?.multiResolve().orEmpty()
-
-                    if (leafSpeck.isStarImport) {
-                        if (mod == resolvedItems.singleOrNull()) {
-                            addAll(crateExportedMacros.value)
-                        }
-                    } else {
-                        val exportedMacros = resolvedItems.filterIsInstance<RsMacro>()
-                            .filter { it in crateExportedMacros.value }
-                        addAll(exportedMacros)
-                    }
-                }
+/**
+ * For given `use foo::{bar, baz::quux, spam::{eggs}}` return the only `("foo", "bar")`.
+ * For wildcard `*` import second element in the pair is null
+ */
+private fun collect2segmentPaths(rootSpeck: RsUseSpeck): List<Pair<String, String?>> {
+    val result = mutableListOf<Pair<String, String?>>()
+
+    fun go(speck: RsUseSpeck, parentPath: RsPath? = null) {
+        testAssert { parentPath?.path == null }
+        val path = speck.path
+        val qualifier = path?.path
+        val group = speck.useGroup
+        if (path != null && qualifier != null && (qualifier.path != null || parentPath != null)) return
+        val firstSegment = parentPath ?: qualifier ?: path
+        val lastSegment = path ?: qualifier ?: parentPath
```

Comment:
This looks quite complicated
I feel it can be simplified with using `RsPath#qualifier` extension method which works with `UseSpeck` too